### PR TITLE
[Core] Enable log rotation for Raylet and GCS server

### PR DESCRIPTION
**Pull request by kfstorm**
## Why are these changes needed?

This PR adds log rotation functionality for Raylet and GCS server by setting a non-empty log_dir when invoking `ray::StartRayLog`. This is critical for production environments to avoid logs taking too much disk space.

After this PR, `raylet_{pid}.log`/`gcs_server_{pid}.log` and potentially `raylet_{pid}.{number}.log`/`gcs_server_{pid}.{number}.log` are created as well as the existing `.out` and `.err` files. See the sample below.

```
dashboard_agent.log
dashboard.log
debug_state_gcs.txt
debug_state.txt
events
gcs_server_309463.log
gcs_server.err
gcs_server.out
log_monitor.log
monitor.err
monitor.log
monitor.out
old
python-core-driver-01000000ffffffffffffffffffffffffffffffffffffffffffffffff_309360.log
python-core-worker-28ccdc393f85a1b96961f0794c9305256c6e02d2d91d2cce1ea76aa6_309765.log
raylet_309662.log
raylet.err
raylet.out
worker-28ccdc393f85a1b96961f0794c9305256c6e02d2d91d2cce1ea76aa6-01000000-309765.err
worker-28ccdc393f85a1b96961f0794c9305256c6e02d2d91d2cce1ea76aa6-01000000-309765.out
```

TODO:
- [x] Code changes in dashboard and agent.
- [x] More test updates and fixes.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
A couple high level comments before deep diving; 
1. We should update this doc. https://docs.ray.io/en/master/ray-observability/ray-logging.html. Can you include changes here?
2. Why not completely removing gcs_server.out and gcs_server.err? We can't because of log monitor?

**Comment by kfstorm** [comment]
> We should update this doc. [docs.ray.io/en/master/ray-observability/ray-logging.html](https://docs.ray.io/en/master/ray-observability/ray-logging.html). Can you include changes here?

No problem.

> Why not completely removing gcs_server.out and gcs_server.err? We can't because of log monitor? (unless we remove it, log rotation has no meaning?)

In some cases, some logs are printed to stderr or stdout without going through `RAY_LOG`. Cases I can think of are:

1. An unhandled C++ exception is thrown.
2. GRPC or ABSL assertion failure.
3. Dashboard agent C++ logs. (Dashboard agent invokes C++ code without initializing Ray C++ logging and it shares stdour and stderr of Raylet.)
4. ERROR or FATAL logs are always printed to stderr as well. See https://github.com/ray-project/ray/issues/12893. I'm not sure if we need to expose Raylet and GCS server error logs to driver as well. If not, we can disable printing to stderr for Raylet and GCS server.

Although stderr and stdout logs are uncommon, I do think they are helpful when debugging issues.

> Why do we need {pid} for raylet? Is it because new raylet can be started on that node again? (isn't this creating a new session name anyway?)

This is the default behavior of Ray C++ logging. In the case of multi-node Python tests, I guess log rotation won't work well because multiple processes share the same log filename pattern. However, I haven't read and tested the internal logic of `rotating_file_sink` of spdlog yet. I can test removing `_{pid}` suffix and see if they work well.

**Comment by kfstorm** [comment]
After reading the implementation of log rotation in spdlog https://github.com/gabime/spdlog/blob/v1.x/include/spdlog/sinks/rotating_file_sink-inl.h, I don't think multiple processes with the same filename pattern will work well.

**Comment by kfstorm** [comment]
@rkooo567 Any more high-level comments before I update the PR?

**Comment by rkooo567** [comment]
Yeah I think it generally looks good. A couple comments;

1. When we have RAY_LOG(ERROR) or RAY_LOG(FATAL), can we also write logs to .err file (so we will have 2 sinks, .log and .err)? i.e. keep the below behavior

> ERROR or FATAL logs are always printed to stderr as well. See https://github.com/ray-project/ray/issues/12893. I'm not sure if we need to expose Raylet and GCS server error logs to driver as well. If not, we can disable printing to stderr for Raylet and GCS server.


2. Right now, if you use cluster_utils.py, it will create raylet.1.out, raylet.2.out, and etc.. Can you just quickly verify how it'd look like after this PR?

I will start the full review soon!

**Comment by rkooo567** [comment]
Also, @wuisawesome @ckw017 would there be any compatibility issue with Loki integration?

**Comment by rkooo567** [comment]
Also for the compatibility (since there are lots of people who will still do cat gcs_server.out), can you write a comment on the top of the .out log file to see .log file instead?

**Comment by kfstorm** [comment]
> 1. When we have RAY_LOG(ERROR) or RAY_LOG(FATAL), can we also write logs to .err file (so we will have 2 sinks, .log and .err)? i.e. keep the below behavior

The behavior is defined in `logging.cc`, and I don't change it. It should apply to Raylet and GCS as well. I added a temporary error log in `main.cc` and here is the result:

```
 ➜  logs grep ' E ' -R .
./raylet_174876.log:[2022-07-07 19:36:16,328 E 174876 174876] main.cc:123: This is a test error log.
./raylet.err:[2022-07-07 19:36:16,132 E 174813 174813] main.cc:123: This is a test error log.
./raylet_174813.log:[2022-07-07 19:36:16,132 E 174813 174813] main.cc:123: This is a test error log.
./raylet.1.err:[2022-07-07 19:36:16,328 E 174876 174876] main.cc:123: This is a test error log.
```

> 2. Right now, if you use cluster_utils.py, it will create raylet.1.out, raylet.2.out, and etc.. Can you just quickly verify how it'd look like after this PR?

This PR won't impact the creation of `.out` and `.err` files. They will be created like before. Below are the log files when I run `python/ray/tests/test_multi_node_2.py::test_cluster`:

```
-rw-r--r-- 1 admin admin   474 Jul  7 19:36 dashboard_agent.log
-rw-r--r-- 1 admin admin  4196 Jul  7 19:36 dashboard.log
-rw-r--r-- 1 admin admin  2344 Jul  7 19:36 debug_state_gcs.txt
-rw-r--r-- 1 admin admin  6138 Jul  7 19:36 debug_state.txt
drwxr-xr-x 2 admin admin  4096 Jul  7 19:36 events
-rw-r--r-- 1 admin admin  7978 Jul  7 19:36 gcs_server_174716.log
-rw-r--r-- 1 admin admin     0 Jul  7 19:36 gcs_server.err
-rw-r--r-- 1 admin admin     0 Jul  7 19:36 gcs_server.out
-rw-r--r-- 1 admin admin   351 Jul  7 19:36 log_monitor.log
-rw-r--r-- 1 admin admin     0 Jul  7 19:36 monitor.err
-rw-r--r-- 1 admin admin  2101 Jul  7 19:36 monitor.log
-rw-r--r-- 1 admin admin     0 Jul  7 19:36 monitor.out
drwxrwxrwx 2 admin admin  4096 Jul  7 19:36 old
-rw-r--r-- 1 admin admin 12881 Jul  7 19:36 raylet_174813.log
-rw-r--r-- 1 admin admin 13163 Jul  7 19:36 raylet_174876.log
-rw-r--r-- 1 admin admin    81 Jul  7 19:36 raylet.1.err
-rw-r--r-- 1 admin admin     0 Jul  7 19:36 raylet.1.out
-rw-r--r-- 1 admin admin    81 Jul  7 19:36 raylet.err
-rw-r--r-- 1 admin admin     0 Jul  7 19:36 raylet.out
```

**Comment by wuisawesome** [comment]
A high level question: why are we doing log rotation like this? (based on pids in this format)? 

My default assumption would've been that we do log rotation apache style.

```
gcs_server.out.2022.06.01
```

**Comment by kfstorm** [comment]
> A high level question: why are we doing log rotation like this? (based on pids in this format)?
> 
> My default assumption would've been that we do log rotation apache style.
> 
> ```
> gcs_server.out.2022.06.01
> ```

This is how spdlog works. See https://github.com/gabime/spdlog/blob/6c95f4c8168401badff89d0c16e3887ea91ea9ad/include/spdlog/sinks/rotating_file_sink-inl.h#L55. It doesn't provide another rotation pattern.

**Comment by rkooo567** [comment]
I think the new semantic lgtm. Just one last qusetion

> The behavior is defined in logging.cc, and I didn't change it. It should apply to Raylet and GCS as well. I added a temporary error log in main.cc and here is the result:

To make sure I understand correctly, we will write all basic stdout to .log file not .out file right? Otherwise, there's no meaning to do log rotation (since .out file will increase forever like before). That said,

.log -> all logs from RAY_LOG() (stdout + stderr)
.out -> stdout output except stdout from RAY_LOG. 
.err -> all stderr output

**Comment by kfstorm** [comment]
> To make sure I understand correctly, we will write all basic stdout to .log file not .out file right? Otherwise, there's no meaning to do log rotation (since .out file will increase forever like before).

Yes, we have to write all `RAY_LOG` logs to .log files because `.out` and `.err` are just stdout and stderr redirection and they don't support log rotation.

**Comment by rkooo567** [comment]
Sounds good. Let's just make it clear from the doc which things are flushed to which file. Let me know when the PR is read y to review

**Comment by kfstorm** [comment]
@rkooo567 This PR is ready for review.

**Comment by rkooo567** [comment]
cc @rickyyx to take a lokot

**Comment by rickyyx** [comment]
> cc @rickyyx to take a look

Sorry - was interrupted by other tasks, will do ASAP.

**Comment by kfstorm** [comment]
@rickyyx Kindly ping.

**Review by rickyyx — COMMENTED** [review]
Hey @kfstorm , so sorry for the delay in reviewing the PR. My apologies for letting this slip. 

It LGTM overall! Thanks for the detailed explanation of before-and-after behaviour of the logging in the comments above, that's really helpful. 

A couple of high level questions I have on top of some nits/questions I commented in-place (mostly docs/elaboration) 
1. IIUC, one of the motives for this PR is to reduce log disk usage, but I am not sure how rotation is helping here? Is there a external service/cron job that I am not aware of that's doing the clean-up? Or was/will that be in another PR?
2. I am not an expert in the specific log library we are using, but is there a way to actually include timestamps info in the filenames other than just "index"? So that people (and other toolings) could locate the specific files easily. 

And please feel free to ping more aggressively for review! :D

**Comment by rickyyx** [comment]
> I am not an expert in the specific log library we are using, but is there a way to actually include timestamps info in the filenames other than just "index"? So that people (and other toolings) could locate the specific files easily.

Ah, I see, the default rotation only supports simple integer indexing. Not sure about your usecases, but I guess most of usecases will not have such long running ray session (other than maybe ray serve), so it should be fine.

**Comment by kfstorm** [comment]
> 1. IIUC, one of the motives for this PR is to reduce log disk usage, but I am not sure how rotation is helping here? Is there a external service/cron job that I am not aware of that's doing the clean-up? Or was/will that be in another PR?

The cleanup is done by the logging framework: spdlog. We don't need to setup cron jobs to clean up log files. I believe that's how most logging frameworks do, not just spdlog.

> 2. I am not an expert in the specific log library we are using, but is there a way to actually include timestamps info in the filenames other than just "index"? So that people (and other toolings) could locate the specific files easily.

> Ah, I see, the default rotation only supports simple integer indexing. Not sure about your usecases, but I guess most of usecases will not have such long running ray session (other than maybe ray serve), so it should be fine.

We've enabled log rotation internally for Core Worker logs of Java workers for more than 2 years, and no one has complained about the index-based naming pattern.

File: python/ray/tests/test_logging.py
Comment (kfstorm) [reply python/ray/tests/test_logging.py:272]: With this PR, logs written by `RAY_LOG` in GCS server and Raylet are written to files by the spdlog logging framework. The env variable `RAY_LOG_TO_STDERR` only disables stderr/stdout redirection, so it won't affect these logs. However, I see that this is only an implementation detail, and we should make sure `RAY_LOG_TO_STDERR` takes effect on all logs. So I think I can disable writing logs to files when `RAY_LOG_TO_STDERR` is true. In this case, spdlog will write logs to stderr by default, and then the logs can be affected by `RAY_LOG_TO_STDERR`.

File: src/ray/util/logging.cc
Comment (kfstorm) [reply src/ray/util/logging.cc:257]: the first parameter of StartRayLog, `app_name`, is `raylet` on macOS/Linux and `raylet.exe` on Windows. If we don't have this special treatment for Windows, the log file name would be `raylet.exe_<pid>.log`. So this part of the code is to remove the `.exe` part from the log file name, to make it consistent with `raylet.out` and `raylet.err` (no `.exe` part even on Windows).

File: src/ray/util/logging.cc
Comment (kfstorm) [reply src/ray/util/logging.cc:307]: What do you mean "the rest of the PRs"? If you mean the test cases, that's because most test cases can't reach the size limit of a single log file. If not, please tell me the exact lines. 😀

File: python/ray/tests/test_logging.py
Comment (kfstorm) [reply python/ray/tests/test_logging.py]: Just found out the behavior is what I described above already. Reverted the test case.

**Comment by rickyyx** [comment]
> The cleanup is done by the logging framework: spdlog. We don't need to setup cron jobs to clean up log files. I believe that's how most logging frameworks do, not just spdlog.

Oh, I see, there's the max number of files parameter. Sorry, wasn't too familiar with the concept of the rotating file. 


> We've enabled log rotation internally for Core Worker logs of Java workers for more than 2 years, and no one has complained about the index-based naming pattern.

I see, that's good to know!

File: src/ray/util/logging.cc
Comment (rickyyx) [reply src/ray/util/logging.cc:307]: Yeah, I guess the fact log rotation never triggered in tests was my concern, i.e. not sure how the system will behave when log rotation kicks in, even though I feel it's 99.9% working :D

 Is there an easy way to actually trigger the log rotation in a test or a manual test run with some steps so that whoever works on it later could reproduce the behavior documented in this PR?

File: python/ray/tests/test_logging.py
Comment (rickyyx) [reply python/ray/tests/test_logging.py]: NIce, thanks!

**Comment by kfstorm** [comment]
@rickyyx Any more comments?

**Comment by rickyyx** [comment]
> @rickyyx Any more comments?

LGTM - cc @rkooo567

**Review by jovany-wang — APPROVED** [review]
I think if there is no more comment from other guys, this is ready to merge.

**Review by rkooo567 — COMMENTED** [review]
Sorry for the delay. I think the code itself is okay, but it is a pretty big backward incompatible change (e.g., it can easily break anyone who scrapes logs now). Let's actually do the API review here, or is it possible to keep the basic log file as raylet.out & gcs_server.out instead of raylet_[pid].log? Potential alternative file name is

```
raylet.out -> same as the current raylet_[pid].log
raylet.err -> same as now

and something else that's for redirecting stdout.
```

**Comment by jovany-wang** [comment]
@rkooo567 
I don't think it's a big backward incompatible change.
1. `raylet.log` and `gcs_server.log` are often used by Ray developers, even most of them are Ray core developers. It's not used widely by our users.

3. We can update the developing guide for how to grep log messages from raylet and gcs server.

**Comment by jovany-wang** [comment]
Maybe CC @scv119 @iycheng @jjyao

**Comment by rkooo567** [comment]
> raylet.log and gcs_server.log are often used by Ray developers, even most of them are Ray core developers. It's not used widely by our users.

Hmm I actually found we explicitly mentioned the file name's backward compatibility is not maintained. So it is probably okay. We still need a doc change (actually I've seen many users reading this log when they debug, and changing the name can break their log scraping depending on how it is implemented). I think we still need an API approval. cc @pcmoritz.

File: doc/source/ray-observability/ray-logging.rst
Comment (rkooo567) [inline doc/source/ray-observability/ray-logging.rst:152]: We need to clearly explain the difference between out / err / pid.log

File: doc/source/ray-observability/ray-logging.rst
Comment (rkooo567) [inline doc/source/ray-observability/ray-logging.rst:159]: same as above

**Comment by jovany-wang** [comment]
> I think we still need an API approval. cc @pcmoritz.
Sure.

File: doc/source/ray-observability/ray-logging.rst
Comment (jovany-wang) [reply doc/source/ray-observability/ray-logging.rst:157]: Good catch~

Comment (jovany-wang) [reply doc/source/ray-observability/ray-logging.rst:158]: DONE. See the latest commit.

**Comment by jovany-wang** [comment]
@ray-project/ray-docs for docs owner approval.

**Comment by jovany-wang** [comment]
@ericl @pcmoritz for the core api approval.

**Pull request closed without merging**
